File: src/gatling/scala/com/example/data/RandomData.scala
package com.example.data

import com.example.models.{Comment, Post, User}
import com.fasterxml.jackson.databind.ObjectMapper
import com.github.javafaker.Faker
import com.google.gson.Gson
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter
import java.util
import java.util.Locale

class RandomData {

  def getProperty(propertyName: String, defaultValue: String): String = {
    Option(System.getenv(propertyName))
      .orElse(Option(System.getProperty(propertyName)))
      .getOrElse(defaultValue)
  }

  val faker = new Faker(new Locale("en_US"))
  val objectMapper = new ObjectMapper()
  val gson = new Gson()
  val isoDateFormat: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
  val simpleDateFormat: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")

  def dateTimeNow(): String = OffsetDateTime.now.format(isoDateFormat)

  def randomUser(): String = {
    objectMapper.writeValueAsString(new User()
      .setName(s"${faker.name().firstName()} ${faker.name().lastName()}")
      .setUsername(faker.name.username())
      .setEmail(faker.internet.emailAddress())
      .setCreatedAt(dateTimeNow()))
  }

  def randomPost(userId: Int): String = {
    def randomCategory(): String = {
      val cats = List("cats", "dogs", "test")
      cats(faker.random().nextInt(0, (cats.size - 1)))
    }
    def randId: Integer = faker.random().nextInt(100, 999)
    objectMapper.writeValueAsString(new Post()
      .setTitle(s"Test Post from ${OffsetDateTime.now()} ${faker.random().nextInt(100000, 999999)}")
      .setSubject("Performance Testing")
      .setBody(faker.lorem().sentence())
      .setUser(Integer.valueOf(userId))
      .setCreatedAt(dateTimeNow())
      .setCategory(randomCategory())
      .setComments(util.Arrays.asList(randId, randId)) //.setComments(util.List.of(faker.random().nextInt(100, 999)))
    )
  }

  def randomComment(postId: Int, email: String): String = {
    objectMapper.writeValueAsString(new Comment()
      .setPost(postId)
      .setEmail(email)
      .setName(s"My Comment ${faker.random().nextInt(100000, 999999)}")
      .setLikes(util.Arrays.asList(0))
      .setDislikes(util.Arrays.asList(0))
      .setCreatedAt(dateTimeNow())
      .setBody("Hello World"))
  }

}

object RandomData extends RandomData {


  override def dateTimeNow(): String = super.dateTimeNow()


}

File: src/gatling/kotlin/com/example/data/RandomData.kt
package com.example.data

import com.example.extensions.toJson
import com.example.modelds.Comment
import com.example.modelds.Post
import com.example.modelds.User
import com.github.javafaker.Faker
import com.google.gson.Gson
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter
import java.util.*

val gson = Gson()
val faker = Faker(Locale.ENGLISH)
fun dateTimeNow(): String = OffsetDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)

val defaultHeaders = mapOf(
    "Content Type" to "application/json; charset=utf8",
    "Accept" to "application/json"
)

fun jsonStr(obj: Any): String = gson.toJson(obj)

/*
Transform array string "[1, 2, 3, 4, 5]" to List<Int>"
 */
fun stringArrToList(arrayString: String): List<Int> = arrayString
    .replace("\\s+","")
    .substring(1, arrayString.length - 1).split(",")
    .map(Integer::valueOf)


fun randomCategory(): String = listOf("cats", "dogs", "test")[faker.random().nextInt(0, 2)]

fun fakeUser(): String =
    User(
        name = "${faker.name().firstName()} ${faker.name().lastName()}",
        username = faker.name().username(),
        email = faker.internet().emailAddress(),
        createdAt = dateTimeNow()
    ).toJson



fun fakePost(userId: Int): String =
    Post(
        title = "My Fake post from ${dateTimeNow()}",
        subject = "Performance Testing",
        body = faker.lorem().sentence(),
        category = randomCategory(),
        user = userId,
        createdAt = dateTimeNow(),
        comments = mutableListOf(faker.random().nextInt(50, 100), faker.random().nextInt(1, 49))
    ).toJson

fun fakeComment(post: Int, email: String): String =
    Comment(
        name = "Positive Comment",
        body = "Hello There",
        likes = mutableListOf(1),
        dislikes = mutableListOf(0),
        email = email,
        post = post,
        createdAt = dateTimeNow(),
    ).toJson

class RandomData {

    companion object {
        val fakeUser: String by lazy {
            User(
                name = "${faker.name().firstName()} ${faker.name().lastName()}",
                username = faker.name().username(),
                email = faker.internet().emailAddress(),
                createdAt = dateTimeNow()
            ).toJson
        }
    }

}
